File: src/services/cryptoService.ts
/**
 * Bank-Grade Crypto Service
 * Uses Web Crypto API for performance and security.
 * Algorithm: AES-256-GCM
 * Key Derivation: HKDF-SHA256
 */

const MASTER_SECRET = process.env.EXPO_PUBLIC_MASTER_SECRET || 'DEFAULT_INSECURE_SECRET_CHANGE_ME_IMMEDIATELY_1234567890';
const ALGORITHM = 'AES-GCM';
const KEY_LENGTH = 256;

// Convert string to Uint8Array
const enc = new TextEncoder();
const dec = new TextDecoder();

async function importMasterKey(): Promise<CryptoKey> {
    return window.crypto.subtle.importKey(
        'raw',
        enc.encode(MASTER_SECRET),
        { name: 'HKDF' },
        false,
        ['deriveKey']
    );
}

async function deriveKey(salt: Uint8Array): Promise<CryptoKey> {
    const masterKey = await importMasterKey();
    return window.crypto.subtle.deriveKey(
        {
            name: 'HKDF',
            hash: 'SHA-256',
            salt: salt,
            info: enc.encode('kiosk-api'),
        },
        masterKey,
        { name: ALGORITHM, length: KEY_LENGTH },
        false,
        ['encrypt', 'decrypt']
    );
}

function arrayBufferToBase64(buffer: ArrayBuffer): string {
    let binary = '';
    const bytes = new Uint8Array(buffer);
    const len = bytes.byteLength;
    for (let i = 0; i < len; i++) {
        binary += String.fromCharCode(bytes[i]);
    }
    return window.btoa(binary);
}

function base64ToArrayBuffer(base64: string): ArrayBuffer {
    const binary_string = window.atob(base64);
    const len = binary_string.length;
    const bytes = new Uint8Array(len);
    for (let i = 0; i < len; i++) {
        bytes[i] = binary_string.charCodeAt(i);
    }
    return bytes.buffer;
}

export const cryptoService = {
    encrypt: async (payload: any, aadStr: string = '') => {
        const plaintext = enc.encode(JSON.stringify(payload));
        const salt = window.crypto.getRandomValues(new Uint8Array(16));
        const iv = window.crypto.getRandomValues(new Uint8Array(12));
        const aad = enc.encode(aadStr);

        const key = await deriveKey(salt);

        const ciphertext = await window.crypto.subtle.encrypt(
            {
                name: ALGORITHM,
                iv: iv,
                additionalData: aad,
            },
            key,
            plaintext
        );

        return {
            iv: arrayBufferToBase64(iv),
            salt: arrayBufferToBase64(salt),
            ciphertext: arrayBufferToBase64(ciphertext),
            aad: aadStr,
        };
    },

    decrypt: async (encryptedFields: { iv: string; salt: string; ciphertext: string; aad: string }) => {
        const iv = new Uint8Array(base64ToArrayBuffer(encryptedFields.iv));
        const salt = new Uint8Array(base64ToArrayBuffer(encryptedFields.salt));
        const ciphertext = base64ToArrayBuffer(encryptedFields.ciphertext);
        const aad = enc.encode(encryptedFields.aad);

        const key = await deriveKey(salt);

        try {
            const decrypted = await window.crypto.subtle.decrypt(
                {
                    name: ALGORITHM,
                    iv: iv,
                    additionalData: aad,
                },
                key,
                ciphertext
            );
            return JSON.parse(dec.decode(decrypted));
        } catch (e) {
            console.error('Decryption failed', e);
            throw new Error('Integrity Check Failed');
        }
    }
};

File: backend/src/main/java/com/kiosk/backend/service/CryptoService.java
package com.kiosk.backend.service;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import javax.crypto.Cipher;
import javax.crypto.Mac;
import javax.crypto.SecretKey;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;
import java.util.Base64;

@Service
public class CryptoService {

    @Value("${kiosk.master.secret:DEFAULT_INSECURE_SECRET_CHANGE_ME_IMMEDIATELY_1234567890}")
    private String masterSecret;

    private static final String ALGORITHM = "AES/GCM/NoPadding";
    private static final int GCM_TAG_LENGTH = 128;
    private static final int SALT_LENGTH = 16;
    private static final int IV_LENGTH = 12;

    private final SecureRandom secureRandom = new SecureRandom();

    /**
     * Decrypts the payload using a key derived from the Master Secret + provided
     * Salt.
     */
    public byte[] decrypt(String ivBase64, String saltBase64, String ciphertextBase64, String aad) {
        try {
            byte[] iv = Base64.getDecoder().decode(ivBase64);
            byte[] salt = Base64.getDecoder().decode(saltBase64);
            byte[] ciphertext = Base64.getDecoder().decode(ciphertextBase64);

            SecretKey derivedKey = deriveKey(salt);

            Cipher cipher = Cipher.getInstance(ALGORITHM);
            GCMParameterSpec spec = new GCMParameterSpec(GCM_TAG_LENGTH, iv);
            cipher.init(Cipher.DECRYPT_MODE, derivedKey, spec);

            if (aad != null) {
                cipher.updateAAD(aad.getBytes(StandardCharsets.UTF_8));
            }

            return cipher.doFinal(ciphertext);
        } catch (Exception e) {
            throw new RuntimeException("Decryption failed: Integrity check or key mismatch", e);
        }
    }

    /**
     * Encrypts the payload, generating a fresh Salt and IV.
     * Returns a wrapper containing all components needed for decryption.
     */
    public EncryptionResult encrypt(byte[] plaintext, String aad) {
        try {
            byte[] salt = new byte[SALT_LENGTH];
            secureRandom.nextBytes(salt);

            byte[] iv = new byte[IV_LENGTH];
            secureRandom.nextBytes(iv);

            SecretKey derivedKey = deriveKey(salt);

            Cipher cipher = Cipher.getInstance(ALGORITHM);
            GCMParameterSpec spec = new GCMParameterSpec(GCM_TAG_LENGTH, iv);
            cipher.init(Cipher.ENCRYPT_MODE, derivedKey, spec);

            if (aad != null) {
                cipher.updateAAD(aad.getBytes(StandardCharsets.UTF_8));
            }

            byte[] ciphertext = cipher.doFinal(plaintext);

            return new EncryptionResult(
                    Base64.getEncoder().encodeToString(iv),
                    Base64.getEncoder().encodeToString(salt),
                    Base64.getEncoder().encodeToString(ciphertext),
                    aad);

        } catch (Exception e) {
            throw new RuntimeException("Encryption failed", e);
        }
    }

    /**
     * HKDF-SHA256 Derivation
     * Derived Key = HKDF(MasterSecret, Salt, Info="kiosk-api")
     */
    private SecretKey deriveKey(byte[] salt) throws Exception {
        byte[] ikm = masterSecret.getBytes(StandardCharsets.UTF_8);
        byte[] info = "kiosk-api".getBytes(StandardCharsets.UTF_8);

        // HKDF-Extract
        Mac hmac = Mac.getInstance("HmacSHA256");
        SecretKeySpec saltKey = new SecretKeySpec(salt, "HmacSHA256"); // Salt is used as key for extract
        hmac.init(saltKey);
        byte[] prk = hmac.doFinal(ikm);

        // HKDF-Expand
        Mac hmacExpand = Mac.getInstance("HmacSHA256");
        hmacExpand.init(new SecretKeySpec(prk, "HmacSHA256"));

        // T(1) = HMAC-Hash(PRK, T(0) | info | 0x01)
        hmacExpand.update(info);
        hmacExpand.update((byte) 0x01);
        byte[] okm = hmacExpand.doFinal(); // 32 bytes for SHA-256 matches AES-256 requirement

        return new SecretKeySpec(okm, "AES");
    }

    public static class EncryptionResult {
        public String iv;
        public String salt;
        public String ciphertext;
        public String aad;

        public EncryptionResult(String iv, String salt, String ciphertext, String aad) {
            this.iv = iv;
            this.salt = salt;
            this.ciphertext = ciphertext;
            this.aad = aad;
        }
    }
}
